dex/farm/src/lib.rs

```diff
@@ -210,6 +221,15 @@ pub trait Farm:
         require!(payment_token_id == farm_token_id, "Bad input token");
         require!(amount > 0, "Payment amount cannot be zero");
 
+        let caller = self.blockchain().get_caller();
```

Comment:
you have to make one function out of lines 224-232. And use a require for it. one liner for this endpoint.

Reply:
I would make a separate version for this contract. And would have especially for this case. You can start a new feat/branch - where you say farm with whitelist.